### DISCUSSION TITLE
Error types: methods or variants?

### DISCUSSION THREAD

**Discussion by clarfonthey** [rust-lang-nursery/api-guidelines#154] [API Guidelines]
Basically: should error types expose what they are via an enum such as `std:io::ErrorKind` or via accessor methods like `reqwest::Error::is_http`?

This seems like there should be some guideline for which is preferred and when.

**Comment by KodrAus** [comment]
I personally prefer methods to variants, because you can introduce subtle runtime bugs by changing the conditions under which a variant is produced. Say we have an error with a `Http` variant:

```rust
#[non_exhaustive]
enum MyError {
    Http,
    OtherVariant,
}
```

Now say we want to introduce a specific `Timeout` variant, which used to be under `Http`:

```rust
#[non_exhaustive]
enum MyError {
    Http,
    Timeout,
    OtherVariant,
}
```

Anyone depending on timeouts caught by a pattern on `Http` won't anymore. But with methods the new timeout error could still return `true` from `is_http`, but a new `is_timeout` method is more specific.
